### PR TITLE
REFACTOR: Simplify controller setup

### PR DESCRIPTION
**Pull request by justindirose**
The code in the plugin needed a dramatic cleanup. This refactor collapses the Plan/Product/Subscription controllers on the backend into one new controller: `SubscribeController`.

This reduces N+1 calls to the back end during the subscription process and simplifies use of the code. 

I've also removed a bunch of dead code and refactored some logic into methods for easier readability. No feature/functionality changes in this PR; only refactoring. However, refactoring will allow for implementation of better anonymous user handling, so this is largely a foundation to enable making that change.

### CODE REVIEW

File: app/controllers/discourse_subscriptions/subscribe_controller.rb [hunk 1]
```diff
@@ -1,31 +1,55 @@
...
+        render_json_dump products
+
+      rescue ::Stripe::InvalidRequestError => e
+        render_json_error e.message
+      end
+    end
+
+    def show
+      begin
+        product = ::Stripe::Product.retrieve(params[:id])
```
Comment (pmusaraj) [inline app/controllers/discourse_subscriptions/subscribe_controller.rb:36]: It wouldn't hurt to add `params.require(:id)` just before this line.

Comment (justindirose) [reply app/controllers/discourse_subscriptions/subscribe_controller.rb:37]: Good catch!

File: app/controllers/discourse_subscriptions/subscribe_controller.rb [hunk 2]
```diff
@@ -1,31 +1,56 @@
...
+
+        render_json_dump response
       rescue ::Stripe::InvalidRequestError => e
         render_json_error e.message
       end
     end
 
     def create
       begin
+        customer = create_customer(params[:source])
```
Comment (pmusaraj) [inline app/controllers/discourse_subscriptions/subscribe_controller.rb:53]: Another case of `params.require`.

File: config/routes.rb [hunk 3]
```diff
@@ -18,14 +18,11 @@
...
   end
 
-  resources :customers, only: [:create]
-  resources :plans, only: [:index], constraints: SubscriptionsUserConstraint.new
-  resources :products, only: [:index, :show]
-  resources :subscriptions, only: [:create]
-
-  post '/subscriptions/finalize' => 'subscriptions#finalize'
+  get '/' => 'subscribe#index'
+  get '.json' => 'subscribe#index'
```
Comment (pmusaraj) [inline config/routes.rb:22]: I don't think this is necessary, the previous route declaration is not limited to a format, so it should respond on `.json` too.

**Review by pmusaraj — APPROVED** [review]
Good-looking cleanup!

File: config/routes.rb (hunk 3 above)
Comment (justindirose) [reply config/routes.rb:22]: Somehow it turned out to be. It works fine in production but tests fail without it. Maybe it's something with how the engine is set up? I'd have to take a look.

File: app/controllers/discourse_subscriptions/subscribe_controller.rb (hunk 2 above)
Comment (justindirose) [reply app/controllers/discourse_subscriptions/subscribe_controller.rb:53]: Good catch!

**Pull request merged**